### DISCUSSION TITLE
Konsist.scopeFromModule()

### DISCUSSION THREAD

**Discussion by timplifier** [LemonAppDev/konsist#562] [Ask a Question]
### Wanted end result.

So I have a structure like that :
feature directory -> authentication directory -> data module
feature directory -> main directory -> data module

I want to scope my Konsist linting for both modules, and maybe do like 
`Konsist.scopeFromModule("feature:authentication:data", ":feature:main:data")` so you have to add module path

because the project name for this module is just `data` regardless the directory it is stored in

### What you tried so far.

it throws 
`com.lemonappdev.konsist.core.exception.KoPreconditionFailedException: Declaration list is empty. Please make sure that list of declarations contain items before calling the 'assert' method`

According to documentation, you have to just do `Konsist.scopeFromModule("data")` in my situation, but it still throws the same error though there are files. This is my test if you want to have a look:

```abstract class ClassAssertionTest {
    abstract val packageClassesResideIn: String

    protected open val classSuffix: String = ""

    protected open val koScope = Konsist.scopeFromProject()

    protected fun assertClassesPackage(
        filter: KoClassDeclaration.() -> Boolean = { name.endsWith(classSuffix) },
    ) {
        koScope.classes()
            .filter { filter(it) }
            .assert { it.resideInPackage(packageClassesResideIn) }
    }

    protected fun assertClassesNamesThatResideInPackage(
        assertion: KoClassDeclaration.() -> Boolean = { name.endsWith(classSuffix) }
    ) {
        koScope.classes().withPackage(packageClassesResideIn)
            .assert { assertion(it) }
    }
}


class RepositoryImplAssertionTest : ClassAssertionTest() {
    override val koScope = Konsist.scopeFromModule("data")
    override val packageClassesResideIn = "..repositories.."

    override val classSuffix = "RepositoryImpl"

    @Test
    fun `ensure that all classes with 'RepositoryImpl' suffix reside in 'repositories' package`() {
        assertClassesPackage()
    }

    @Test
    fun `ensure that all classes that reside in 'repositories' package have 'RepositoryImpl' suffix`() {
        assertClassesNamesThatResideInPackage()
    }
}

**Comment by igorwojda** [comment]
Nested modules are not fully supported yet.
We have a ticket in the backlog to address this at some point.

**Reply by timplifier** [reply]
so what is the solution you may suggest ?

**Reply by igorwojda** [reply]
You can use `Konsist.scopeFromDirectory("feature/authentication/data")` for now
https://docs.konsist.lemonappdev.com/writing-tests/koscope#directory-scope

**Comment by igorwojda** [comment]
We have some community feedback that `scopeFromModule` may work as well. 
See updated docs https://docs.konsist.lemonappdev.com/writing-tests/koscope#nested-module-scope

Let me know if this works for you

BTW The nested modules feature is not complete. The community is reporting that this feature works, however, we still have to take a closer look, review expectations, and add tests.
